src/main/scala/com.snowplowanalytics/sqs2kinesis/EventsStreamModule.scala

```diff
@@ -58,89 +55,133 @@ object EventsStreamModule {
+  def kinesisFlow(config: Sqs2KinesisConfig, kinesisClient: AmazonKinesisAsync): Flow[Vector[ParsedMsg], Message, NotUsed] = {
+
+    // The innter flow, which must be retried on error.
+    val inner = Flow[(Vector[ParsedMsg], Vector[ParsedMsg])]
+      .map { case (todo, complete) =>
+        val req = new PutRecordsRequest()
+          .withStreamName(config.kinesisStreamName)
+          .withRecords(todo.map(toPutRecordReqEntry).asJavaCollection)
+        try {
+          val results = kinesisClient.putRecords(req).getRecords.asScala.toList
+          val (successes, failures) = todo.zip(results).partition(_._2.getErrorMessage == null)
+          (complete ++ successes.map(_._1), failures.map(_._1))
+        } catch {
+          case NonFatal(e) =>
+            logger.error("Writing to kinesis failed with error", e)
+            (complete, todo)
+        }
       }
 
-    val toPutRecordReqEntry: KinesisKeyAndMsg => PutRecordsRequestEntry = {
-      case (key, data) =>
-        new PutRecordsRequestEntry().withPartitionKey(key).withData(data)
-    }
+    Flow[Vector[ParsedMsg]]
+      .map(_ -> Vector.empty)
+      .via(RetryFlow.withBackoff(500.milli, 1.second, 0.0, 5, inner) {
+        case (_, (_, Vector())) =>
+          None
+        case (_, (successes, failures)) =>
+          if (failures.nonEmpty)
+            logger.error(s"Got ${failures.size} failures and ${successes.size} successes writing to kinesis. Failures will be retried.")
+          Some((failures, successes))
+      }).mapConcat { case (successes, failures) =>
+        if (failures.nonEmpty)
+          logger.error(s"Got ${failures.size} failures and ${successes.size} successes writing to kinesis. Giving up on failures because max retries exceeded. Failures will not be acked to sqs.")
+        logger.debug(s"Successfully wrote ${successes.size} messages to kinesis")
+        successes.map(_.original)
+      }
+  }
+
+  /** A Flow that tries to ack a sqs message. Upon failure, the ack will be retried up to 5 times */
+  def confirmSqsSink(config: Sqs2KinesisConfig)(implicit client: SqsAsyncClient): Sink[Message, NotUsed] = {
+    val inner = Flow[Message]
+      .map(MessageAction.Delete(_))
+      .via(SqsAckFlow(config.sqsQueue))
+      .map(_ => Option.empty[Throwable])
+      .recover {
+        case NonFatal(e) => Some(e)
+      }
 
-    def stackTraceToString(e: Throwable): String = {
-      val out = new ByteArrayOutputStream()
-      val ps  = new PrintStream(out)
-      e.printStackTrace(ps)
-      val stack = out.toString
-      ps.close
-      out.close
-      stack
+    RetryFlow.withBackoff(500.milli, 1.second, 0.0, 5, inner) {
+      case (in, Some(e)) =>
+        logger.warn("Error acking sqs message. It will be retried", e)
+        Some(in)
+      case (_, None) => None
     }
+    .to(Sink.foreach {
+      case Some(e) => logger.error("Exceeded retry limit acking sqs", e)
```

Comment:
If this happens, we should assume sqs is unhealthy, or there's loss of network, or some other io error.

After logging the error, we give up trying to ack the message.  Eventually our lease on the sqs message times out, so sqs sends it to us again, 30 seconds later (assuming sqs or the network has recovered!)  We receive the message again and write it to kinesis again, so there will be a duplicate.

It's a close call, but I think it is slightly better to stay alive in this situation.  Our app is healthy still, even if sqs is unhealthy.  Even if we crash the app, it will get re-started again immediately by whatever orchestration platform is managing the service.  So that's not much different from just staying alive and retrying.  Typically, there would be monitoring on the sqs queue size so an alarm would go off if messages are not being consumed.